# RetailVision AI
## Notebook 02 - Data Cleaning & Feature Engineering

### Objective
Prepare the raw retail sales data for analysis by performing data cleaning, handling missing values, merging datasets, and creating new features.

### Expected Output
A clean master dataset ready for Exploratory Data Analysis (EDA), SQL, Power BI, and Machine Learning.

In [1]:
import pandas as pd
import numpy as 

In [2]:
train = pd.read_csv(r"C:\Users\Saumya S\OneDrive\Desktop\RetailVision-AI\dataset\train.csv", low_memory=False)

store = pd.read_csv(r"C:\Users\Saumya S\OneDrive\Desktop\RetailVision-AI\dataset\store.csv")

In [3]:
train["Date"].dtype

dtype('O')

In [5]:
train["Date"] = pd.to_datetime(train["Date"])

In [6]:
train["Date"].dtype

dtype('<M8[ns]')

In [7]:
store.isnull().sum()

Store                          0
StoreType                      0
Assortment                     0
CompetitionDistance            3
CompetitionOpenSinceMonth    354
CompetitionOpenSinceYear     354
Promo2                         0
Promo2SinceWeek              544
Promo2SinceYear              544
PromoInterval                544
dtype: int64

In [8]:
store["CompetitionDistance"] = store["CompetitionDistance"].fillna(
    store["CompetitionDistance"].median()
)

In [9]:
store.isnull().sum()

Store                          0
StoreType                      0
Assortment                     0
CompetitionDistance            0
CompetitionOpenSinceMonth    354
CompetitionOpenSinceYear     354
Promo2                         0
Promo2SinceWeek              544
Promo2SinceYear              544
PromoInterval                544
dtype: int64

In [12]:
data = train.merge(store , on = "Store", how = "left")

In [13]:
data.shape

(1017209, 18)

In [14]:
data.head()

,Store,DayOfWeek,Date,Sales,Customers,Open,Promo,StateHoliday,SchoolHoliday,StoreType,Assortment,CompetitionDistance,CompetitionOpenSinceMonth,CompetitionOpenSinceYear,Promo2,Promo2SinceWeek,Promo2SinceYear,PromoInterval
0,1,5,2015-07-31,5263,555,1,1,0,1,c,a,1270.0,9.0,2008.0,0,NaN,NaN,NaN
1,2,5,2015-07-31,6064,625,1,1,0,1,a,a,570.0,11.0,2007.0,1,13.0,2010.0,"Jan,Apr,Jul,Oct"
2,3,5,2015-07-31,8314,821,1,1,0,1,a,a,14130.0,12.0,2006.0,1,14.0,2011.0,"Jan,Apr,Jul,Oct"
3,4,5,2015-07-31,13995,1498,1,1,0,1,c,c,620.0,9.0,2009.0,0,NaN,NaN,NaN
4,5,5,2015-07-31,4822,559,1,1,0,1,a,a,29910.0,4.0,2015.0,0,NaN,NaN,NaN


In [15]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1017209 entries, 0 to 1017208
Data columns (total 18 columns):
 #   Column                     Non-Null Count    Dtype         
---  ------                     --------------    -----         
 0   Store                      1017209 non-null  int64         
 1   DayOfWeek                  1017209 non-null  int64         
 2   Date                       1017209 non-null  datetime64[ns]
 3   Sales                      1017209 non-null  int64         
 4   Customers                  1017209 non-null  int64         
 5   Open                       1017209 non-null  int64         
 6   Promo                      1017209 non-null  int64         
 7   StateHoliday               1017209 non-null  object        
 8   SchoolHoliday              1017209 non-null  int64         
 9   StoreType                  1017209 non-null  object        
 10  Assortment                 1017209 non-null  object        
 11  CompetitionDistance        1017209 no

# 4. Feature Engineering

## Objective

The `Date` column contains complete date information, but it is not very useful for business analysis in its current form.

In this step, we will extract useful features from the Date column, such as Year, Month, Day, Weekday, and Quarter.

These new features will help answer business questions like:

- Which month has the highest sales?
- Which weekday generates the most revenue?
- Which quarter performs the best?

These features will also improve the performance of our Machine Learning model.

In [23]:
# Extract Year
data["Year"] = data["Date"].dt.year

# Extract Month
data["Month"] = data["Date"].dt.month

# Extract Day
data["Day"] = data["Date"].dt.day

# Extract Weekday
data["Weekday"] = data["Date"].dt.day_name()

# Extract Quarter
data["Quarter"] = data["Date"].dt.quarter

In [24]:
data[["Date","Year","Month","Day","Weekday","Quarter"]].head()

,Date,Year,Month,Day,Weekday,Quarter
0,2015-07-31,2015,7,31,Friday,3
1,2015-07-31,2015,7,31,Friday,3
2,2015-07-31,2015,7,31,Friday,3
3,2015-07-31,2015,7,31,Friday,3
4,2015-07-31,2015,7,31,Friday,3


## Observations

The Date column has been successfully transformed into multiple useful features.

These features will be used throughout the project for:

- Exploratory Data Analysis (EDA)
- SQL Business Queries
- Power BI Dashboards
- Machine Learning Models

Instead of working with a single Date column, we can now analyze sales trends by year, month, weekday, and quarter.

In [25]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1017209 entries, 0 to 1017208
Data columns (total 23 columns):
 #   Column                     Non-Null Count    Dtype         
---  ------                     --------------    -----         
 0   Store                      1017209 non-null  int64         
 1   DayOfWeek                  1017209 non-null  int64         
 2   Date                       1017209 non-null  datetime64[ns]
 3   Sales                      1017209 non-null  int64         
 4   Customers                  1017209 non-null  int64         
 5   Open                       1017209 non-null  int64         
 6   Promo                      1017209 non-null  int64         
 7   StateHoliday               1017209 non-null  object        
 8   SchoolHoliday              1017209 non-null  int64         
 9   StoreType                  1017209 non-null  object        
 10  Assortment                 1017209 non-null  object        
 11  CompetitionDistance        1017209 no

In [26]:
data.to_csv(
    r"C:\Users\Saumya S\OneDrive\Desktop\RetailVision-AI\dataset\cleaned_data.csv",
    index=False
)

# Final Observations

The dataset has been successfully prepared for analysis.

### Data Cleaning Summary

- Converted the `Date` column to datetime format.
- Filled missing values in `CompetitionDistance` using the median.
- Merged `train.csv` and `store.csv` into a single master dataset.
- Created five new time-based features:
  - Year
  - Month
  - Day
  - Weekday
  - Quarter

### Dataset Status

- Total Records: **1,017,209**
- Total Features: **23**
- Duplicate Rows: **0**
- Missing values only exist where they are expected from the business logic (Promo2-related columns).

The dataset is now ready for Exploratory Data Analysis (EDA), SQL analysis, Power BI visualization, and Machine Learning.